system/lib/wasmfs/wasmfs.cpp

```diff
@@ -70,14 +70,18 @@ WasmFS::~WasmFS() {
   rootDirectory->locked().setParent(nullptr);
 }
 
-std::shared_ptr<Directory> WasmFS::initRootDirectory() {
-
+// Special backends that want to install themselves as the root use this hook.
+// Otherwise, we use the default backends.
+__attribute__((weak)) extern backend_t wasmfs_create_root_dir(void) {
```

Comment:
Let's declare this in the wasmfs.h hooks section to make it available to users who want to e.g. use OPFS as their root directory.

Reply:
Do we want this to be public? I'm not sure myself. I feel like if a user tries to use it at the same time as a backend there could be a weird failure, whereas in-tree, we control all the uses (so we can error on a flag that conflicts with NODERAWFS for example). But maybe I'm overly worried?

Reply:
I think it would be nice to make available to users. Backends in general should not be injecting themselves as the root directories, since that's not composable at all (since only one of them can win). I would view this instead as the _application_ choosing what the root directory should be. The NODERAWFS linker flag is one mechanism the application can use to control its root directory, but I don't see any harm in giving it more general control via this hook.

Maybe we should rename noderawfs_backend.cpp to something else to not pretend it's a proper backend. Maybe `noderawfs_root.cpp`?

Reply:
Good point! Yeah, it's not really a backend. I think it only was a backend in the old FS for technical reasons. Yes, renaming to `_root` sgtm.